File: android_app/src/main/java/me/timeto/app/ui/PrivacySheet.kt
package me.timeto.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import me.timeto.app.*
import me.timeto.shared.vm.PrivacySheetVM

private val hPadding = MyListView.PADDING_OUTER_HORIZONTAL

@Composable
fun PrivacySheet(
    layer: WrapperView.Layer,
) {

    val (vm, state) = rememberVM { PrivacySheetVM() }

    VStack(
        modifier = Modifier
            .fillMaxHeight()
            .background(c.sheetBg),
    ) {

        val scrollState = rememberScrollState()

        Sheet__HeaderView(
            title = state.headerTitle,
            scrollState = scrollState,
            bgColor = c.sheetBg,
        )

        VStack(
            modifier = Modifier
                .verticalScroll(state = scrollState)
                .padding(bottom = 20.dp)
                .weight(1f),
        ) {

            PView(state.text1, topPadding = 8.dp)

            PView(state.text2)

            PView(state.text3)

            PView(state.text4, fontWeight = FontWeight.Bold)

            PView(state.text5)

            VStack(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 16.dp)
                    .background(c.sheetFg)
                    .padding(top = 8.dp, bottom = 12.dp),
            ) {
                state.sendItems.forEach { sendItem ->
                    PView(sendItem, topPadding = 4.dp)
                }
            }

            PView(state.text6)

            MyListView__ItemView(
                isFirst = true,
                isLast = true,
                modifier = Modifier
                    .padding(top = 24.dp),
                withTopDivider = false,
            ) {
                MyListView__ItemView__SwitchView(
                    text = state.sendReportsTitle,
                    isActive = state.isSendReportsEnabled,
                ) {
                    vm.toggleIsSendingReports()
                }
            }

            Text(
                text = "Open Source",
                modifier = Modifier
                    .padding(top = 16.dp)
                    .padding(horizontal = MyListView.PADDING_INNER_HORIZONTAL)
                    .clip(squircleShape)
                    .clickable {
                        showOpenSource()
                    }
                    .padding(horizontal = 8.dp, vertical = 4.dp),
                color = c.blue,
                fontSize = 14.sp,
            )
        }

        Sheet__BottomViewClose {
            layer.close()
        }
    }
}

@Composable
private fun PView(
    text: String,
    topPadding: Dp = 16.dp,
    fontWeight: FontWeight = FontWeight.Normal,
) {
    Text(
        text = text,
        modifier = Modifier
            .padding(horizontal = hPadding)
            .padding(top = topPadding),
        color = c.white,
        lineHeight = 22.sp,
        fontWeight = fontWeight,
    )
}

File: ios_app/iphone/PrivacySheet.swift
import SwiftUI
import shared

private let hPadding = MyListView.PADDING_OUTER_HORIZONTAL

struct PrivacySheet: View {

    @Binding var isPresented: Bool

    @State private var vm = PrivacySheetVM()
    @State private var scroll = 0

    var body: some View {

        VMView(vm: vm, stack: .VStack()) { state in

            Sheet__HeaderView(
                title: state.headerTitle,
                scrollToHeader: scroll,
                bgColor: c.sheetBg
            )

            ScrollViewWithVListener(showsIndicators: false, vScroll: $scroll) {

                VStack {

                    PView(text: state.text1, topPadding: 8)

                    PView(text: state.text2)

                    PView(text: state.text3)

                    PView(text: state.text4, fontWeight: .bold)

                    PView(text: state.text5)

                    VStack {
                        ForEachIndexed(state.sendItems) { _, sendItem in
                            PView(text: sendItem, topPadding: 4)
                        }
                    }
                            .padding(.top, 8)
                            .padding(.bottom, 12)
                            .background(c.sheetFg)
                            .padding(.top, 16)

                    PView(text: state.text6)

                    /*
                MyListView__ItemView(
                    isFirst = true,
                    isLast = true,
                    modifier = Modifier
                            .padding(top = 24.dp),
                    withTopDivider = false,
                    ) {
                    MyListView__ItemView__SwitchView(
                        text = state.sendReportsTitle,
                        isActive = state.isSendReportsEnabled,
                        ) {
                        vm.toggleIsSendingReports()
                    }
                }
            */

                    Button(
                        action: {
                            showOpenSource()
                        },
                        label: {
                            HStack {

                                Text("Open Source")
                                    .padding(.top, 16)
                                    .padding(.horizontal, MyListView.PADDING_OUTER_HORIZONTAL)
                                    .padding(.top, 8)
                                    .foregroundColor(c.blue)
                                    .font(.system(size: 15))

                                Spacer()
                            }
                        }
                    )
                }
                        .padding(.bottom, 20)
            }

            Sheet__BottomViewClose {
                isPresented = false
            }
        }
    }
}

private struct PView: View {

    let text: String
    var topPadding: CGFloat = 16
    var fontWeight: Font.Weight = .regular

    var body: some View {

        Text(text)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, hPadding)
                .padding(.top, topPadding)
                .foregroundColor(c.white)
                .font(.system(size: 17, weight: fontWeight))
                .multilineTextAlignment(.leading)
                .lineSpacing(2)
    }
}
